### PR TITLE
ODT: Zone odt refacto

### PR DESCRIPTION
**Pull request by antoine-de**
For zone odt we want to change the current behavior.

We store the admin with zone odt, and we a journey is asked we add the odt stop_points of the admin if they exists.

That way we'll have a 0mn fallback to the ODT station (which is the reality since the taxi comes to pick you at your door)

Fixes http://jira.canaltp.fr/browse/NAVITIAII-1077

### CODE REVIEW

File: source/type/data.cpp [hunk 1]
```diff
@@ -426,6 +427,26 @@ void Data::build_odt(){
...
     }
+
+    //for zone odt, we weed to fill the Admin structure
+    for (const auto jp: pt_data->journey_patterns) {
+        if (jp->odt_level != type::OdtLevel_e::zonal) {
+            continue;
+        }
+
+        if (jp->journey_pattern_point_list.size() != 2) {
+            LOG4CPLUS_INFO(log4cplus::Logger::getInstance("log"), "it's strange, a zone odt journey pattern ("
```
Comment (kinnou02) [inline source/type/data.cpp]: Warn maybe

Comment (antoine-de) [inline source/type/data.cpp]: I though we wanted warn for important stuff, but yes I can change the level (or maybe info for the log, and warn for the number of pb if >0)

**Comment by antoine-de** [comment]
Jenkins, test this please

File: source/routing/raptor_api.cpp [hunk 2]
```diff
@@ -509,31 +544,78 @@ static void add_isochrone_response(RAPTOR& raptor,
...
+    return data.geo_ref->find_admins(ep.coordinates);
+}
+
+static
+std::vector<std::pair<type::idx_t, navitia::time_duration> >
+get_stop_points( const type::EntryPoint &ep, const type::Data& data,
+        georef::StreetNetwork & worker, bool use_second = false){
     std::vector<std::pair<type::idx_t, navitia::time_duration> > result;
     log4cplus::Logger logger = log4cplus::Logger::getInstance(LOG4CPLUS_TEXT("logger"));
     LOG4CPLUS_DEBUG(logger, "Searching nearest stop_point's from entry point : [" << ep.coordinates.lat()
```
Comment (kinnou02) [inline source/routing/raptor_api.cpp]: maybe it's time to delete this log, or a least set the level at trace

File: source/type/data.cpp [hunk 3]
```diff
@@ -413,6 +414,49 @@ void Data::aggregate_odt(){
...
+            for (auto* admin: sp->admin_list) {
+                odt_stops_by_admin[admin].insert(sp);
+            }
+        }
+    }
+
+    //we first store the stops in a set not to have dupplicates
+    for (const auto& p: odt_stops_by_admin) {
+        for (const auto& sp: p.second) {
+            std::cout << "on rajoute: " << sp->uri << " a l'admin " << p.first->uri << std::endl;
```
Comment (kinnou02) [inline source/type/data.cpp]: debug

**Comment by antoine-de** [comment]
Jenkins, test this please

File: source/ed/connectors/fusio_parser.h [hunk 4]
```diff
@@ -283,7 +283,7 @@ struct CalendarLineFusioHandler : public GenericHandler {
 };
 }
 
-struct AdminStopAreaFusioHandler : public GenericHandler {
+struct  AdminStopAreaFusioHandler : public GenericHandler {
```
Comment (TeXitoi) [inline source/ed/connectors/fusio_parser.h]: Are you shure?

File: source/routing/raptor_api.cpp [hunk 5]
```diff
@@ -151,9 +151,26 @@ static void add_pathes(EnhancedResponse& enhanced_response,
...
-        if (!path.items.front().stop_points.empty()
+        const nt::VehicleJourney* first_vj = path.items.front().get_vj();
+        const nt::VehicleJourney* last_vj = path.items.back().get_vj();
+        bool journey_begin_with_address_odt =
+                (first_vj && in(first_vj->vehicle_journey_type,
+        {nt::VehicleJourneyType::adress_to_stop_point, nt::VehicleJourneyType::odt_point_to_point}));
+
+        bool journey_end_with_address_odt =
+                (last_vj && in(last_vj->vehicle_journey_type,
+        {nt::VehicleJourneyType::adress_to_stop_point, nt::VehicleJourneyType::odt_point_to_point}));
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: Strange indentation

File: source/routing/raptor_api.cpp [hunk 6]
```diff
@@ -151,9 +151,26 @@ static void add_pathes(EnhancedResponse& enhanced_response,
...
+         *     (since the stop point has been artificially created on the data side)
+         *     we want a odt section from the departure asked by the user to the destination of the odt)
          **/
 
-        if (!path.items.front().stop_points.empty()
+        const nt::VehicleJourney* first_vj = path.items.front().get_vj();
+        const nt::VehicleJourney* last_vj = path.items.back().get_vj();
+        bool journey_begin_with_address_odt =
+                (first_vj && in(first_vj->vehicle_journey_type,
+        {nt::VehicleJourneyType::adress_to_stop_point, nt::VehicleJourneyType::odt_point_to_point}));
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: Strange indentation

File: source/routing/raptor_api.cpp [hunk 7]
```diff
@@ -151,9 +151,26 @@ static void add_pathes(EnhancedResponse& enhanced_response,
...
+         * 4) 'taxi like' odt. For some case, we don't want a walking section to the the odt stop point
+         *     (since the stop point has been artificially created on the data side)
+         *     we want a odt section from the departure asked by the user to the destination of the odt)
          **/
 
-        if (!path.items.front().stop_points.empty()
+        const nt::VehicleJourney* first_vj = path.items.front().get_vj();
+        const nt::VehicleJourney* last_vj = path.items.back().get_vj();
+        bool journey_begin_with_address_odt =
+                (first_vj && in(first_vj->vehicle_journey_type,
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: first_vj can be nullptr?

File: source/routing/raptor_api.cpp [hunk 8]
```diff
@@ -151,9 +151,26 @@ static void add_pathes(EnhancedResponse& enhanced_response,
...
+        {nt::VehicleJourneyType::adress_to_stop_point, nt::VehicleJourneyType::odt_point_to_point}));
+
+        bool journey_end_with_address_odt =
+                (last_vj && in(last_vj->vehicle_journey_type,
+        {nt::VehicleJourneyType::adress_to_stop_point, nt::VehicleJourneyType::odt_point_to_point}));
+
+        if (journey_begin_with_address_odt) {
+            //first is zonal ODT, we do nothing, there is no walking nor crow fly section, but we'll have to update the start of the journey
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: long line

File: source/routing/raptor_api.cpp [hunk 9]
```diff
@@ -509,31 +544,78 @@ static void add_isochrone_response(RAPTOR& raptor,
...
-                 bool use_second = false) {
+template <typename T>
+const std::vector<georef::Admin*>& get_admins(const std::string& uri, const std::unordered_map<std::string, T*>& obj_map) {
+    auto it = obj_map.find(uri);
+    if (it == obj_map.end()) {
+        static const std::vector<georef::Admin*> empty_list;
+        return empty_list;
+    }
+    return it->second->admin_list;
+}
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp:557]: This is exactly `find_or_default` (well, not exactly, but almost)

File: source/routing/raptor_api.cpp [hunk 10]
```diff
@@ -509,31 +544,78 @@ static void add_isochrone_response(RAPTOR& raptor,
...
+
+const std::vector<georef::Admin*> find_admins(const type::EntryPoint& ep, const type::Data& data) {
+    if (ep.type == type::Type_e::StopArea) {
+        return get_admins<type::StopArea>(ep.uri, data.pt_data->stop_areas_map);
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: explicit template parameter is useless

File: source/routing/raptor_api.cpp [hunk 11]
```diff
@@ -509,31 +544,78 @@ static void add_isochrone_response(RAPTOR& raptor,
...
+const std::vector<georef::Admin*>& get_admins(const std::string& uri, const std::unordered_map<std::string, T*>& obj_map) {
+    auto it = obj_map.find(uri);
+    if (it == obj_map.end()) {
+        static const std::vector<georef::Admin*> empty_list;
+        return empty_list;
+    }
+    return it->second->admin_list;
+}
+
+const std::vector<georef::Admin*> find_admins(const type::EntryPoint& ep, const type::Data& data) {
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: const is useless or you forgot a reference. If you opt for the reference, be careful, returning a new object will be buggy

File: source/routing/raptor_api.cpp [hunk 12]
```diff
@@ -509,31 +544,78 @@ static void add_isochrone_response(RAPTOR& raptor,
     }
 }
 
-static std::vector<std::pair<type::idx_t, navitia::time_duration> >
-get_stop_points( const type::EntryPoint& ep,
-                 const type::Data& data,
-                 georef::StreetNetwork& worker,
-                 bool use_second = false) {
+template <typename T>
+const std::vector<georef::Admin*>& get_admins(const std::string& uri, const std::unordered_map<std::string, T*>& obj_map) {
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp:550]: need static

File: source/routing/raptor_api.cpp (hunk 11 above)
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: need static

File: source/type/data.cpp [hunk 13]
```diff
@@ -413,6 +414,48 @@ void Data::aggregate_odt(){
...
+            for (auto* admin: sp->admin_list) {
+                odt_stops_by_admin[admin].insert(sp);
+            }
+        }
+    }
+
+    //we first store the stops in a set not to have dupplicates
+    for (const auto& p: odt_stops_by_admin) {
+        for (const auto& sp: p.second) {
+            p.first->odt_stop_points.push_back(sp);
```
Comment (TeXitoi) [inline source/type/data.cpp:456]: odt_stop_points order will not be determisistic, is it a problem?

File: source/routing/raptor_api.cpp (hunk 7 above)
Comment (antoine-de) [inline source/routing/raptor_api.cpp]: I think so

File: source/type/data.cpp (hunk 13 above)
Comment (antoine-de) [inline source/type/data.cpp:456]: I don't think so, It is only used  to have free stoppoints on those admin, the order is irrelevant.

**Pull request merged**